### DISCUSSION TITLE
Might build_ext.py support pypi version of swig?

### DISCUSSION THREAD

**Discussion by ChaselChiu** [pypa/setuptools#4544] [Q&A]
I'm new to this, but looks like currently build_ext.py only support either swig.exe or swig linux-style app, but with "pip install swig", we should be able to "import swig" and call "swig._program("swig", argv)" regardless windows or linux. Could we support it?

  **current model:**
     . "apt install swig (linux)" or "download swig.zip-> unzip-> add-to-path (windows)"
     . pip install pylibfdt (will fail without swig)
  **desired new model with pypi swig support:**
     . pip install swig (both linux and windows)
     . pip install pylibfdt
If I misunderstood something please help to correct me. Or if you have any other idea which can simplify the process please help to share too.

Thanks,
Chasel

Relevant code:
https://github.com/pypa/setuptools/blob/main/setuptools/_distutils/command/build_ext.py#L633

**Comment by abravalheri** [comment]
If `swig` is distributed via PyPi, installing it already places the executable in the PATH, right?

If that is so, then I don't think there is much else to do.

If not, then it would be interesting to discuss with the `swig` package on PyPI how to automatically put the executable in the PATH.

**Reply by ChaselChiu** [reply]
Thanks for response!
Just my guess, it might be safer that pip packages do not modify system environment variables (like path variable), so it looks to me they handle the path within python script scope:
https://github.com/nightlark/swig-pypi/blob/main/src/swig/__init__.py

**Reply by abravalheri** [reply]
`pip` does not modify the PATH variable directly. If I understood correctly, this Python package has a console script entry-point. So what pip does is to create a brand new executable that is placed into the `bin` or `Scripts` designated to `pip` according to the relevant installation directory schema (that changes if it is a `--user` instalation,  `virtualenv`, etc...).

**Reply by abravalheri** [reply]
`pip` does not modify the PATH variable directly. If I understood correctly, this Python package has a console script entry-point. So what pip does is to create a brand new executable that is placed into the `bin` or `Scripts` designated to `pip` according to the relevant installation directory schema (that changes if it is a `--user` instalation,  `virtualenv`, etc...).

If your environment was set up accordingly,  this folder is already part of the PATH.

**Reply by abravalheri** [reply]
Anyway, if you are looking to have a specific package built with that version of `swig`, you need to follow the documentation and have the build dependency on `swig` listed in the `pyproject.toml` file of the package you are trying to build.

**Reply by ChaselChiu** [reply]
Yes, we are trying pyproject.toml, but current pylibfdt looks relying on setuptool to call swig. I think we need more time to study and investigate.
Thanks for all the data you shared, that's very helpful!

**Comment by ChaselChiu** [comment]
Hello, just FYI that we have co-worked with pylibfdt owners to enhance the package, now we do not need to manually install swig, now the single line command "pip install pylibfdt" is good enough. Thanks for your support and clarification.
